[#59391] rework file links to paginated collection

Kharonus opened the pull request:
# Ticket
[OP#59391](https://community.openproject.org/wp/59391)

# What are you trying to accomplish?
- show again a correct counter on the files tab in work package detail view

# What approach did you choose and why?
- files tab counter now only calls for file links of wp with page size 0
- file links endpoint returns paginated collections
- frontend requests file links with page size -1 (INFINITE)
- file sync is still executed for ALL file links of a work package, not only the requested page
  - this is practically no change for the product, as we do not fetch single pages
  - but fetching only the total numbers with page size 0 now does not trigger a sync

Kharonus committed:
[#59391] rework file links to paginated collection

- https://community.openproject.org/wp/59391
- files tab counter now only calls for file links of wp with page size 0
- file links endpoint returns paginated collections
- frontend requests file links with page size -1 (INFINITE)
- file sync is still executed for ALL file links of a work package, not
  only the requested page
  - this is practically no change for the product, as we do not fetch
    single pages
  - but fetching only the total numbers with page size 0 now does not
    trigger a sync

Kharonus committed:
[#59391] updated API specs for new pagination in file links

- avoid sql injection
- file link collection is returned ordered by id asc

brunopagno reviewed (approved):
Code changes look good to me. I am assuming the pagination change is considered a non breaking API change and thus can be included without any issues to consumers.

:tada:

Kharonus commented:
@brunopagno This is an important point. My take on it would be, that usage of it is in the most cases the same. If you do not define a pagesize, you get a standard back. only if you have more then 30 file links in one collection, you suddenly won't get all back.

This could be considered breaking, but the solution for it is quite simple and needs only adding a `pageSize=-1` to the request's query params.

apfohl commented on modules/storages/lib/api/v3/file_links/join_origin_status_to_file_links_relation.rb:
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
```
You might add the copyright thingy

Kharonus replied:
done

apfohl commented on modules/storages/lib/api/v3/file_links/join_origin_status_to_file_links_relation.rb:
```diff
@@ -0,0 +1,31 @@
+        # @param [Hash] id_status_map A hash mapping file link IDs to their origin status
+        # in the format { 137: "view_allowed", 142: "error" }
+        def self.create(id_status_map)
+          sanitized_sql = ActiveRecord::Base.send(
```
Do you need to use `send`? Can't you just call the method? `sanitize_sql_array([...])`

Kharonus replied:
Uh, interesting. I took this code sample from rubocop's improvements. Will try it.

Kharonus replied:
works, good catch!

apfohl commented on modules/storages/lib/api/v3/file_links/work_packages_file_links_api.rb:
```diff
@@ -28,50 +28,78 @@
+            sync_result = ::Storages::FileLinkSyncService
+                            .new(user: current_user)
+                            .call(file_links)
+                            .result
```
Will it always succeed?

Kharonus replied:
The sync service indeed is always successful. While single storage syncs can fail, the sync service only accumulates successes into its own result.

brunopagno commented:
> This could be considered breaking, but the solution for it is quite simple and needs only adding a `pageSize=-1` to the request's query params.

@Kharonus the behaviour change _could_ be considered breaking. Do we have any previous events similar to this to take into account? Would be nice if the decision was made for us :sweat_smile: 

I personally do not feel the change is a big deal, but if any consumer depends on this behaviour they could be annoyed, even if the fix on the client side is simple.

Kharonus committed:
[#59391] applied PR comments

Kharonus commented:
Well, TBH in the past we did not really bothered with breaking changes to the API. We renamed properties, changed JSON representations, and still released them without a deprecation round.

For this thing here a deprecation round would be adding this behavior behind some kind of feature flag.

It is possible, but it would raise of course again the question, if the change is breaking enough. It should not happen, that all custom clients suddenly break. And for those, where big collections are an issue, the fix is rather simple.

brunopagno commented:
@Kharonus sounds good enough to me. Thanks for the context there :bow:
